src/dbnode/storage/index.go

```diff
@@ -1560,17 +1561,47 @@ func (i *nsIndex) queryWithSpan(
 		state = asyncQueryExecState{
 			exhaustive: true,
 		}
-		deadline = start.Add(timeout)
-		wg       sync.WaitGroup
+		deadline      = start.Add(timeout)
+		wg            sync.WaitGroup
+		totalWaitTime time.Duration
 	)
 
 	// Create a cancellable lifetime and cancel it at end of this method so that
 	// no child async task modifies the result after this method returns.
 	cancellable := xresource.NewCancellableLifetime()
+	// always cancel to cleanup. this might be a no-op if it was already canceled by the background task to interrupt
+	// the workers.
 	defer cancellable.Cancel()
 
-	indexMatchingStartTime := time.Now()
-	var totalWaitTime time.Duration
+	stdCtx, ctxFound := ctx.GoContext()
+	if !ctxFound {
+		stdCtx = stdctx.Background()
+	}
```

Comment:
Should we add a convenience method to the M3 context such as:
```go
goCtx := ctx.GoContextOrBackground()
```

Reply:
I went with `ctx.MustGoContext() stdctx.Context` since it is guaranteed to be set for a thrift rpc handler. Also added a new helper constructors for tests `NewBackground()` and `NewWithGoContext()`